File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineCollector.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collector;
import java.util.stream.Stream;

class LogLineCollector {

    private final Predicate<String> isCommitSeparator;

    private LogLineCollector(Predicate<String> isCommitSeparator) {
        this.isCommitSeparator = isCommitSeparator;
    }

    static Collector<String, ?, Stream<List<String>>> create(Predicate<String> commitSeparatorTest) {
        LogLineCollector collector = new LogLineCollector(commitSeparatorTest);
        return Collector.of(ArrayList::new, collector::collectLogLine, collector::combineForParallelExecution, collector::removeIncompleteCommits);
    }

    private void collectLogLine(List<List<String>> commits, String logLine) {
        if (isCommitSeparator.test(logLine)) {
            startNewCommit(commits);
        } else {
            assertOneCommitIsPresent(commits);
            addToLastCommit(commits, logLine);
        }
    }

    private void startNewCommit(List<List<String>> commits) {
        commits.add(new ArrayList<>());
    }

    private void assertOneCommitIsPresent(List<List<String>> commits) {
        if (commits.isEmpty()) {
            throw new IllegalArgumentException("no commit present or parallel stream of log lines, which is not supported");
        }
    }

    private void addToLastCommit(List<List<String>> commits, String logLine) {
        int indexOfLastCommit = commits.size() - 1;
        List<String> lastCommit = commits.get(indexOfLastCommit);
        lastCommit.add(logLine);
    }

    private List<List<String>> combineForParallelExecution(List<List<String>> firstList, List<List<String>> secondList) {
        throw new UnsupportedOperationException("parallel collection of log lines not supported");
    }

    private Stream<List<String>> removeIncompleteCommits(List<List<String>> commits) {
        return commits.stream().filter(commit -> !commit.isEmpty());
    }

}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineCollector.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser

import java.util.function.Function
import java.util.function.BiConsumer
import java.util.function.BinaryOperator
import java.util.function.Predicate
import java.util.function.Supplier
import java.util.stream.Collector
import java.util.stream.Stream

class LogLineCollector private constructor(private val isCommitSeparator: Predicate<String>) {

    private val BOM = "\uFEFF"

    private fun collectLogLine(commits: MutableList<MutableList<String>>, logLine: String) {
        val sanitizedLogLine = sanitizeLogLine(logLine)
        if (isCommitSeparator.test(sanitizedLogLine)) {
            startNewCommit(commits)
        } else {
            assertOneCommitIsPresent(commits)
            addToLastCommit(commits, sanitizedLogLine)
        }
    }

    private fun sanitizeLogLine(logLine: String): String {
        if (logLine.startsWith(BOM)) return logLine.substring(1)
        return logLine
    }

    private fun startNewCommit(commits: MutableList<MutableList<String>>) {
        commits.add(ArrayList())
    }

    private fun assertOneCommitIsPresent(commits: List<List<String>>) {
        if (commits.isEmpty()) {
            throw IllegalArgumentException("no commit present, unsupported file encoding, or parallel stream of log lines")
        }
    }

    private fun addToLastCommit(commits: MutableList<MutableList<String>>, logLine: String) {
        val indexOfLastCommit = commits.size - 1
        val lastCommit = commits[indexOfLastCommit]
        lastCommit.add(logLine)
    }

    private fun combineForParallelExecution(
        firstList: MutableList<MutableList<String>>,
        secondList: MutableList<MutableList<String>>
    ): MutableList<MutableList<String>> {
        throw UnsupportedOperationException("parallel collection of log lines not supported")
    }

    private fun removeIncompleteCommits(commits: MutableList<MutableList<String>>): Stream<MutableList<String>> {
        return commits.stream().filter { commit -> !commit.isEmpty() }
    }

    companion object {

        fun create(commitSeparatorTest: Predicate<String>): Collector<String, *, Stream<List<String>>> {
            val collector = LogLineCollector(commitSeparatorTest)
            return Collector.of<String, MutableList<MutableList<String>>, Stream<List<String>>>(
                    Supplier<MutableList<MutableList<String>>> { ArrayList() },
                    BiConsumer<MutableList<MutableList<String>>, String> { commits, logLine ->
                        collector.collectLogLine(commits, logLine)
                    },
                    BinaryOperator<MutableList<MutableList<String>>> { firstList, secondList ->
                        collector.combineForParallelExecution(firstList, secondList)
                    },
                    Function<MutableList<MutableList<String>>, Stream<List<String>>> {
                        collector.removeIncompleteCommits(it).map { it.toList() }
                    }
            )
        }
    }
}
